classes/opensesameapi.php

```diff
@@ -42,106 +42,124 @@
+        $clientid = $this->clientid;
+        $clientsecret = $this->clientsecret;
         $this->setHeader([
                 'Content-Type: application/x-www-form-urlencoded',
                 'Accept: application/json',
                 sprintf('Authorization: Basic %s', base64_encode(sprintf('%s:%s', $clientid, $clientsecret)))
         ]);
-
+        $authurl = $this->authurl;
         $response = $this->post($authurl, 'grant_type=client_credentials&scope=content'
         );
 
         $decoded = json_decode($response);
-        $token = $decoded->access_token;
-        set_config('bearertoken', $token, 'tool_opensesame');
-        set_config('bearertokencreatetime', time(), 'tool_opensesame');
-        $createtime = get_config('tool_opensesame', 'bearertokencreatetime');
-        set_config('bearertokenexpiretime', ($createtime + $decoded->expires_in), 'tool_opensesame');
-        return $this->get_auth_token();
+        if (!$decoded->access_token) {
+            mtrace('Access token is missing reattempting authentication process.');
+            if (!isset($this->access_token)) {
+                throw new \moodle_exception( 'access_token_missing', 'tool_opensesame');
+            }
+            return false;
+        } else {
+            mtrace('Store the access token for later retrieval.');
+            $this->access_token = $decoded->access_token;
+            $this->courserequesturl = $this->createcourserequesturl(); // Create course request url.
+            return true;
+        }
     }
 
     /**
-     * get_auth_token validates token not expired, if expired, creates a new one.
+     * Create the course request url.
      *
-     * @return false|mixed|object|string|null $token
-     * @throws \dml_exception
+     * @return string
      */
-    public function get_auth_token() {
-        mtrace('get_auth_token called');
-        $token = get_config('tool_opensesame', 'bearertoken');
-        $expiretime = get_config('tool_opensesame', 'bearertokenexpiretime');
-        $now = time();
-
-        if ($token === '' || $now >= $expiretime) {
-            mtrace('Token either does not exist or is expired. Token is being created');
-            return  $this->authenticate();
-        } else if ($token !== '' && $now <= $expiretime) {
-            mtrace('Token is valid.');
-            // Define url for the next function.
-            $url = get_config('tool_opensesame', 'baseurl') . '/v1/content?customerIntegrationId=' .
-                    get_config('tool_opensesame', 'customerintegrationid') . '&limit=10';
-            $this->get_open_sesame_course_list($token, $url);
-
-            return $token;
-        }
+    protected function createcourserequesturl():string {
+        mtrace('Creating course request url');
+        $baseurl = $this->baseurl;
+        $customerintegrationid = $this->customerintegrationid;
+        return "{$baseurl}/v1/content?customerIntegrationId={$customerintegrationid}&limit=10";
```

Comment:
Why only retrieve 10 courses per call?

Reply:
This was just because it was easier to test on a limited number of courses for testing purposes only. I have moved it up to 50. However we can move this to whatever number we mutually agree will be best for the server.